### PR TITLE
PublicApi oauth Template, service account, endpoints for Public API

### PR DESCRIPTION
**Pull request by mkralik3**
//skip-ci

Created template and service account for PublicApi
Created endpoints for PublicApi

### CODE REVIEW

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

File: rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+@Slf4j
+public class EnvironmentsPublicApiSteps {
+
+    @Autowired
+    private EnvironmentsPublicEndpoint environmentsEndpoint;
+
+    @When("check that number of tags is (\\w+)$")
+    public void checkNumberOfTags(int numberOfTags) throws IOException {
+        List<String> list = getAllTags();
+        assertThat(list.size()).isEqualTo(numberOfTags);
```
Comment (avano) [inline rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java]: I'd say that the `list` variable and `getAllTags()` method can be left out and 

```
assertThat(environmentsEndpoint.getAllEnvironments().size()).isEqualTo(numberOfTags)
```

should be enough. wdyt?

File: rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    @When("update tag with name ([^\"]*) to ([^\"]*)$")
+    public void updateTag(String tagName, String newTagName) {
+        environmentsEndpoint.renameEnvironment(tagName, newTagName);
+    }
+
+    @When("delete tag with name ([^\"]*)$")
+    public void deleteTag(String tagName) {
+        environmentsEndpoint.deleteEnvironment(tagName);
+    }
+}
```
Comment (avano) [inline rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java]: add an empty new line

File: rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java (hunk 1 above)
Comment (avano) [reply rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java]: and all methods that are checking something should be `@Then`

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/ConnectionsPublicEndpoint.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        ObjectMapper mapper = new ObjectMapper();
+        JsonNode propertiesJson = mapper.convertValue(properties, JsonNode.class);
+
+        Response response = invocation.post(Entity.entity(propertiesJson, MediaType.APPLICATION_JSON));
+        if (response.getStatus() != HttpURLConnection.HTTP_OK) {
+            fail("Status of the response is " + response.getStatus() + ". The response is " + response.getStatusInfo());
+        }
+        return response.readEntity(ConnectionOverview.class);
+    }
+}
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/ConnectionsPublicEndpoint.java]: add a newline

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/EnvironmentsPublicEndpoint.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    public EnvironmentsPublicEndpoint() {
+        super("/environments");
+    }
+
+    /**
+     * Get all tags in the syndesis
+     * endpoint -> GET ​/public​/environments
+     * original method -> {@link io.syndesis.server.endpoint.v1.handler.external.PublicApiHandler#getReleaseEnvironments()}
+     */
+    public List<String> getAllEnvironments() throws IOException {
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/EnvironmentsPublicEndpoint.java]: rather than throwing IOException it would be better to catch it here and fail the test (and then all methods that use `getAllEnvironments` do not need to handle it)

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/EnvironmentsPublicEndpoint.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+     * original method -> {@link io.syndesis.server.endpoint.v1.handler.external.PublicApiHandler#renameEnvironment(String, String)}
+     */
+    public void renameEnvironment(String environment, String newEnvironment) {
+        Invocation.Builder invocation = this.createInvocation(getWholeUrl(String.format(rootEndPoint + "/%s", environment)));
+        Response response = invocation.put(Entity.entity(newEnvironment, MediaType.APPLICATION_JSON));
+        if (response.getStatus() != HttpURLConnection.HTTP_NO_CONTENT) {  //404 #gh-
+            fail("Status of the response is " + response.getStatus() + ". The response is " + response);
+        }
+    }
+}
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/EnvironmentsPublicEndpoint.java]: add a newline

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/EnvironmentsPublicEndpoint.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * Rename particular tag.
+     * endpoint -> PUT /public​/environments​/{env}
+     * original method -> {@link io.syndesis.server.endpoint.v1.handler.external.PublicApiHandler#renameEnvironment(String, String)}
+     */
+    public void renameEnvironment(String environment, String newEnvironment) {
+        Invocation.Builder invocation = this.createInvocation(getWholeUrl(String.format(rootEndPoint + "/%s", environment)));
+        Response response = invocation.put(Entity.entity(newEnvironment, MediaType.APPLICATION_JSON));
+        if (response.getStatus() != HttpURLConnection.HTTP_NO_CONTENT) {  //404 #gh-
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/EnvironmentsPublicEndpoint.java]: there is something missing in the comment :yum:

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/IntegrationsPublicEndpoint.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import javax.ws.rs.client.Entity;
+import javax.ws.rs.client.Invocation;
+import javax.ws.rs.core.GenericType;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import javax.ws.rs.core.SecurityContext;
+import java.io.*;
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/IntegrationsPublicEndpoint.java]: avoid star imports

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/IntegrationsPublicEndpoint.java [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+     * @param pathToZipFile - path to zip file (e.g. /tmp/export.zip)
+     */
+    public void importIntegration(String tag, String pathToZipFile) {
+        MultipartFormDataOutput mdo = new MultipartFormDataOutput();
+        try {
+            mdo.addFormData("data", new FileInputStream(new File(pathToZipFile)),
+                    MediaType.APPLICATION_OCTET_STREAM_TYPE);
+            mdo.addFormData("environment", tag, MediaType.TEXT_PLAIN_TYPE);
+        } catch (FileNotFoundException e) {
+            fail("IO Exception during loading ZIP file.");
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/IntegrationsPublicEndpoint.java]: use `fail("msg", e)` so that you know what the exception is about

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/IntegrationsPublicEndpoint.java [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+     * @param all        - when all is set to true, all integrations are exported and tagged with the tag.
+     */
+    public void exportIntegration(String tag, String pathToSave, boolean all) {
+        Invocation.Builder invocation = this.createInvocation(getWholeUrl(String.format(rootEndPoint + "/%s/export.zip?all=%s", tag, all)));
+        Response res = invocation.get();
+        InputStream is = res.readEntity(InputStream.class);
+        try {
+            exportZip(is, pathToSave);
+        } catch (IOException e) {
+            fail("IO Exception during exporting ZIP file.");
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/IntegrationsPublicEndpoint.java]: `fail("msg", e)`

Comment (avano) [reply utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/IntegrationsPublicEndpoint.java]: or better fail the test in `exportZip` method

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/IntegrationsPublicEndpoint.java [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+     */
+    private void exportZip(InputStream is, String pathToSave) throws IOException {
+        byte[] buffer = new byte[is.available()];
+        is.read(buffer);
+        File targetFile = new File(pathToSave);
+        OutputStream outStream = new FileOutputStream(targetFile);
+        outStream.write(buffer);
+        outStream.close();
+    }
+}
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/IntegrationsPublicEndpoint.java]: add a newline

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/PublicEndpoint.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+    String getWholeUrl(String publicEndpointUrl) {
+        return String.format("https://%s%s%s", PublicOauthProxyTemplate.PUBLIC_API_PROXY_ROUTE, apiPath, publicEndpointUrl);
+    }
+
+    Invocation.Builder createInvocation(String url) {
+        log.info(String.format("Creating invocation for url %s", url));
+        return client.target(url)
+                .request(MediaType.APPLICATION_JSON)
+                .headers(COMMON_HEADERS);
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/PublicEndpoint.java]: I think this can be refactored into RestUtils directly so that we'll avoid duplicating the same code as in AbstractEndpoint. Something like the `client` would be in the `RestUtils` class and it will also have the `createInvocation` method.. but if you prefer we can log a jira for that and do it in separate PR

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/PublicEndpoint.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    }
+}
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/PublicEndpoint.java]: add a newline here

File: utilities/src/main/java/io/syndesis/qe/utils/PublicApiUtils.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+    private PublicApiUtils() {
+    }
+
+    /**
+     * Method creates service account for Public API and gives them competent role
+     */
+    public static void createServiceAccount() {
+        Optional<ServiceAccount> existServiceAccount = OpenShiftUtils.client().serviceAccounts().list().getItems().stream()
+                .filter(t -> t.getMetadata().getName().equalsIgnoreCase(SERVICE_ACCOUNT_NAME))
+                .findFirst();
```
Comment (avano) [inline utilities/src/main/java/io/syndesis/qe/utils/PublicApiUtils.java]: when you know the name, you don't need to iterate over the list and you can use `OpenshiftUtils.client().serviceAccounts().withName("").get()`

File: rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java (hunk 2 above)
Comment (mkralik3) [reply rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java]: I don't know why my Idea doing this, and checkstyle doesn't warn me about it. I have to change something in Settings for this empty new line and star imports.

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/EnvironmentsPublicEndpoint.java (hunk 6 above)
Comment (mkralik3) [reply utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/EnvironmentsPublicEndpoint.java]: It shouldn't be there :(

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/EnvironmentsPublicEndpoint.java (hunk 4 above)
Comment (mkralik3) [reply utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/EnvironmentsPublicEndpoint.java]: In the new commit with the REST tests, I use `response.readEntity(new GenericType<List<String>>() {})` without IOException. I update it here because I have not already done REST tests

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/PublicEndpoint.java (hunk 11 above)
Comment (mkralik3) [reply utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/PublicEndpoint.java]: Yes, it can be refactored in some way. However, the AbstractEndpoint and PublicEndpoint are differently (PublicEndpoint uses Authorization token, design of AbstractEndpoint cannot be applied on PublicApi, ) and now I don't know how to unify it without huge refactoring. And it should cover all class which uses Client.
e.g.
We can create some class (ClientManager ?) which will manage Client instance as common headers, creating an invocation with specific url ane specific media type etc. And this class (ClientManager ?) will use all classes (AbstractEndpoint, PublicEndpoint, ApiProviderSteps, CustomApiEndpoint, Verifier ...)

Comment (avano) [reply utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/PublicEndpoint.java]: something like that I had in my mind as well.. I'll create a jira for that

File: rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java (hunk 2 above)
Comment (avano) [reply rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java]: I'll talk to tomas p. if we should tighten the checkstyle a bit, because currently we have the same configuration as engineering and it has only a very basic set of rules

File: utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/PublicEndpoint.java (hunk 11 above)
Comment (mkralik3) [reply utilities/src/main/java/io/syndesis/qe/endpoints/publicendpoint/PublicEndpoint.java]: ok, after REST tests I can look at it. In my POV, we can skip it for this PR.

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED

File: rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    @Then("check that tag with name ([^\"]*) is not in the tag list$")
+    public void checkTagNotExist(String tagName) {
+        List<String> list = getAllTags();
+        assertThat(list).doesNotContain(tagName);
+    }
+
+    private List<String> getAllTags() {
+        return environmentsEndpoint.getAllEnvironments();
```
Comment (avano) [inline rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java]: currently this method seems to be unnecessary

File: rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    @Then("check that tag with name ([^\"]*) is in the tag list$")
+    public void checkTagExist(String tagName) {
+        List<String> list = getAllTags();
+        assertThat(list).contains(tagName);
+    }
+
+    @Then("check that tag with name ([^\"]*) is not in the tag list$")
+    public void checkTagNotExist(String tagName) {
+        List<String> list = getAllTags();
+        assertThat(list).doesNotContain(tagName);
```
Comment (avano) [inline rest-tests/src/test/java/io/syndesis/qe/rest/tests/publicapi/EnvironmentsPublicApiSteps.java]: refactor these two methods to `assertThat(getAllTags())...` or even better to `assertThat(environmentsEndpoint.getAllEnvironments())...` if the `getAllTags()` is not necessary

**Comment by syndesisqe** [comment]
Build finished. Tests PASSED
